File: backend/services/collection_service.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from lib.mongodb import get_database
from models.collection import Collection, CollectionPermission
from models.user import User, UserRole
from bson import ObjectId

class CollectionService:
    def __init__(self):
        self._db = None
        self._collections_collection = None
    
    @property
    def db(self):
        if self._db is None:
            self._db = get_database()
        return self._db
    
    @property
    def collections_collection(self):
        if self._collections_collection is None:
            self._collections_collection = self.db.collections
        return self._collections_collection

    async def get_all_collections(self) -> List[Collection]:
        """Get all collections"""
        try:
            collections_cursor = self.collections_collection.find({})
            collections = []
            async for collection_doc in collections_cursor:
                collection_doc["id"] = str(collection_doc.pop("_id"))
                collections.append(Collection(**collection_doc))
            return collections
        except Exception as e:
            print(f"Error getting all collections: {e}")
            return []

    async def get_collection_by_id(self, collection_id: str) -> Optional[Collection]:
        """Get collection by ID"""
        try:
            collection_doc = await self.collections_collection.find_one({"_id": ObjectId(collection_id)})
            if collection_doc:
                collection_doc["id"] = str(collection_doc.pop("_id"))
                return Collection(**collection_doc)
            return None
        except Exception as e:
            print(f"Error getting collection by ID: {e}")
            return None

    async def create_collection(self, name: str, permission: CollectionPermission, created_by: User, model_id: Optional[str] = None) -> Collection:
        """Create a new collection"""
        try:
            # Validate collection name
            if not name or not name.strip():
                raise ValueError("Collection name cannot be empty")
            
            name = name.strip()
            if len(name) < 3:
                raise ValueError("Collection name must be at least 3 characters long")
            
            if len(name) > 100:
                raise ValueError("Collection name cannot exceed 100 characters")
            
            # Check for invalid characters (only allow alphanumeric, spaces, hyphens, underscores)
            import re
            if not re.match(r'^[a-zA-Z0-9\s\-_]+$', name):
                raise ValueError("Collection name can only contain letters, numbers, spaces, hyphens, and underscores")
            
            # Check for duplicate name
            existing_collection = await self.collections_collection.find_one({"name": name})
            if existing_collection:
                raise ValueError(f"Collection with name '{name}' already exists")
            
            collection_doc = {
                "_id": ObjectId(),
                "name": name,
                "permission": permission,  # Remove .value since permission is already a string
                "createdBy": created_by.username,
                "createdAt": datetime.utcnow().isoformat()
            }
            
            # Only add modelId if it's provided
            if model_id:
                collection_doc["modelId"] = model_id
            
            result = await self.collections_collection.insert_one(collection_doc)
            collection_doc["id"] = str(collection_doc.pop("_id"))
            return Collection(**collection_doc)
        except Exception as e:
            print(f"Error creating collection: {e}")
            raise ValueError(f"Failed to create collection: {str(e)}")

    async def update_collection(self, collection_id: str, updates: Dict[str, Any]) -> Optional[Collection]:
        """Update collection"""
        try:
            # Remove None values
            updates = {k: v for k, v in updates.items() if v is not None}
            if not updates:
                return await self.get_collection_by_id(collection_id)
            
            result = await self.collections_collection.find_one_and_update(
                {"_id": ObjectId(collection_id)},
                {"$set": updates},
                return_document=True
            )
            
            if result:
                result["id"] = str(result.pop("_id"))
                return Collection(**result)
            return None
        except Exception as e:
            print(f"Error updating collection: {e}")
            return None

    async def delete_collection(self, collection: Collection, current_user: User) -> bool:
        """Delete collection"""
        try:
            result = await self.collections_collection.delete_one({"_id": ObjectId(collection.id)})
            return result.deleted_count > 0
        except Exception as e:
            print(f"Error deleting collection: {e}")
            return False

    def can_user_access_collection(self, user: User, collection: Collection) -> bool:
        """Check if user can access collection"""
        # Public collections can be accessed by anyone
        if collection.permission == "public":
            return True
        
        # Private collections can only be accessed by creator
        if collection.permission == "private":
            return collection.createdBy == user.username
        
        # Department collections can be accessed by same department users
        if collection.permission == "department":
            return (collection.createdBy == user.username or 
                   user.role in [UserRole.ADMIN, UserRole.SUPER_ADMIN])
        
        # Default to no access
        return False

# Create singleton instance
collection_service = CollectionService() 

File: backend/src/services/collectionService.ts
import { Collection, ICollection, CollectionPermission } from '../models/collection';
import { BaseService } from './base/BaseService';
import { ServiceResult } from '../types/base';

export class CollectionService extends BaseService<ICollection> {
  constructor() {
    super(Collection);
    console.log('✅ Collection service initialized');
  }

  /**
   * Get all collections with user-specific filtering
   */
  async getAllCollections(userId?: string, department?: string): Promise<ServiceResult<ICollection[]>> {
    try {
      let filter: any = {};

      if (userId) {
        filter = {
          $or: [
            { createdBy: userId },
            { permission: CollectionPermission.PUBLIC },
            { 
              permission: CollectionPermission.DEPARTMENT,
              department: department 
            }
          ]
        };
      }

      return await this.findWithFilter(filter);
    } catch (error) {
      return {
        success: false,
        error: `Failed to fetch collections: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Create collection with validation
   */
  async createCollection(collectionData: Partial<ICollection>): Promise<ServiceResult<ICollection>> {
    try {
      // Validation ชื่อ collection ตาม legacy
      if (!collectionData.name || typeof collectionData.name !== 'string' || !collectionData.name.trim()) {
        return {
          success: false,
          error: 'Collection name cannot be empty'
        };
      }
      const name = collectionData.name.trim();
      if (name.length < 3) {
        return {
          success: false,
          error: 'Collection name must be at least 3 characters long'
        };
      }
      if (name.length > 100) {
        return {
          success: false,
          error: 'Collection name cannot exceed 100 characters'
        };
      }
      if (!/^[a-zA-Z0-9\s\-_]+$/.test(name)) {
        return {
          success: false,
          error: 'Collection name can only contain letters, numbers, spaces, hyphens, and underscores'
        };
      }
      // Check if collection name already exists (case-insensitive)
      const existingCollection = await this.model.findOne({ name: { $regex: `^${name}$`, $options: 'i' } }).exec();
      if (existingCollection) {
        return {
          success: false,
          error: 'Collection name already exists'
        };
      }

      const collection = new Collection({
        ...collectionData,
        createdAt: new Date(),
        updatedAt: new Date()
      });

      const savedCollection = await collection.save();
      console.log(`✅ Created collection: ${savedCollection.name}`);
      
      return {
        success: true,
        data: savedCollection,
        message: 'Collection created successfully'
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to create collection: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Update collection with ownership check
   */
  async updateCollection(collectionId: string, updates: Partial<ICollection>, userId?: string): Promise<ServiceResult<ICollection | null>> {
    try {
      // Check ownership/role if userId provided
      if (userId) {
        const existingCollection = await this.model.findById(collectionId);
        if (!existingCollection) {
          return {
            success: false,
            error: 'Collection not found'
          };
        }
        // ดึง user object
        const userService = require('./userService');
        const user = await userService.userService.get_user_by_id(userId);
        if (!user) {
          return {
            success: false,
            error: 'User not found'
          };
        }
        const isAdmin = user.role === 'Admin' || user.role === 'SuperAdmin';
        // Logic ตาม legacy
        if (existingCollection.permission === 'PRIVATE') {
          if (existingCollection.createdBy !== user.username && !isAdmin) {
            return {
              success: false,
              error: 'You can only update your own collections'
            };
          }
        } else if (existingCollection.permission === 'PUBLIC') {
          if (!isAdmin) {
            return {
              success: false,
              error: 'Only admin or superadmin can update public collections'
            };
          }
        } else if (existingCollection.permission === 'DEPARTMENT') {
          if (existingCollection.createdBy !== user.username && !isAdmin) {
            return {
              success: false,
              error: 'You can only update your own department collections or be admin/superadmin'
            };
          }
        }
      }
      // Validation ชื่อ collection ตาม legacy
      if (updates.name) {
        const name = updates.name.trim();
        if (!name) {
          return {
            success: false,
            error: 'Collection name cannot be empty'
          };
        }
        if (name.length < 3) {
          return {
            success: false,
            error: 'Collection name must be at least 3 characters long'
          };
        }
        if (name.length > 100) {
          return {
            success: false,
            error: 'Collection name cannot exceed 100 characters'
          };
        }
        if (!/^[a-zA-Z0-9\s\-_]+$/.test(name)) {
          return {
            success: false,
            error: 'Collection name can only contain letters, numbers, spaces, hyphens, and underscores'
          };
        }
        // Check if new name conflicts with existing collection (case-insensitive)
        const existingCollection = await this.model.findOne({ name: { $regex: `^${name}$`, $options: 'i' }, _id: { $ne: collectionId } }).exec();
        if (existingCollection) {
          return {
            success: false,
            error: 'Collection name already exists'
          };
        }
      }

      updates.updatedAt = new Date();
      return await super.update(collectionId, updates);
    } catch (error) {
      return {
        success: false,
        error: `Failed to update collection: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Delete collection with ownership check
   */
  async deleteCollection(collectionId: string, userId?: string): Promise<ServiceResult<boolean>> {
    try {
      // Check ownership/role if userId provided
      if (userId) {
        const existingCollection = await this.model.findById(collectionId);
        if (!existingCollection) {
          return {
            success: false,
            error: 'Collection not found'
          };
        }
        // ดึง user object
        const userService = require('./userService');
        const user = await userService.userService.get_user_by_id(userId);
        if (!user) {
          return {
            success: false,
            error: 'User not found'
          };
        }
        const isAdmin = user.role === 'Admin' || user.role === 'SuperAdmin';
        // Logic ตาม legacy
        if (existingCollection.permission === 'PRIVATE') {
          if (existingCollection.createdBy !== user.username && !isAdmin) {
            return {
              success: false,
              error: 'You can only delete your own collections'
            };
          }
        } else if (existingCollection.permission === 'PUBLIC') {
          if (!isAdmin) {
            return {
              success: false,
              error: 'Only admin or superadmin can delete public collections'
            };
          }
        } else if (existingCollection.permission === 'DEPARTMENT') {
          if (existingCollection.createdBy !== user.username && !isAdmin) {
            return {
              success: false,
              error: 'You can only delete your own department collections or be admin/superadmin'
            };
          }
        }
      }

      return await super.delete(collectionId);
    } catch (error) {
      return {
        success: false,
        error: `Failed to delete collection: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Find collection by name
   */
  async findByName(name: string): Promise<ServiceResult<ICollection | null>> {
    try {
      const collection = await this.model.findOne({ name }).exec();
      
      if (!collection) {
        return {
          success: false,
          error: 'Collection not found'
        };
      }

      return {
        success: true,
        data: collection
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to find collection: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Get collections by permission
   */
  async getCollectionsByPermission(permission: CollectionPermission): Promise<ServiceResult<ICollection[]>> {
    try {
      const collections = await this.model.find({ permission }).exec();
      
      return {
        success: true,
        data: collections,
        message: `Found ${collections.length} collections with permission ${permission}`
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to fetch collections: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Get collections by department
   */
  async getCollectionsByDepartment(department: string): Promise<ServiceResult<ICollection[]>> {
    try {
      const collections = await this.model.find({ department }).exec();
      
      return {
        success: true,
        data: collections,
        message: `Found ${collections.length} collections in department ${department}`
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to fetch collections: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Search collections
   */
  async searchCollections(query: string, userId?: string, department?: string): Promise<ServiceResult<ICollection[]>> {
    try {
      const searchRegex = new RegExp(query, 'i');
      let filter: any = {
        name: searchRegex
      };

      if (userId) {
        filter = {
          $and: [
            { name: searchRegex },
            {
              $or: [
                { createdBy: userId },
                { permission: CollectionPermission.PUBLIC },
                { 
                  permission: CollectionPermission.DEPARTMENT,
                  department: department 
                }
              ]
            }
          ]
        };
      }

      const collections = await this.model.find(filter).exec();
      
      return {
        success: true,
        data: collections,
        message: `Found ${collections.length} collections matching "${query}"`
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to search collections: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Get collections created by user
   */
  async getCollectionsByUser(userId: string): Promise<ServiceResult<ICollection[]>> {
    try {
      const collections = await this.model.find({ createdBy: userId }).exec();
      
      return {
        success: true,
        data: collections,
        message: `Found ${collections.length} collections created by user`
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to fetch user collections: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Get collection statistics
   */
  async getCollectionStats(): Promise<ServiceResult<any>> {
    try {
      const [totalCollections, publicCount, privateCount, departmentCount] = await Promise.all([
        this.model.countDocuments(),
        this.model.countDocuments({ permission: CollectionPermission.PUBLIC }),
        this.model.countDocuments({ permission: CollectionPermission.PRIVATE }),
        this.model.countDocuments({ permission: CollectionPermission.DEPARTMENT })
      ]);

      const stats = {
        total: totalCollections,
        byPermission: {
          public: publicCount,
          private: privateCount,
          department: departmentCount
        }
      };

      return {
        success: true,
        data: stats,
        message: 'Collection statistics retrieved successfully'
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to get collection statistics: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }

  /**
   * Check if user has access to collection
   */
  async hasAccess(collectionId: string, userId?: string, department?: string): Promise<ServiceResult<boolean>> {
    try {
      const collection = await this.model.findById(collectionId).exec();
      if (!collection) {
        return {
          success: false,
          error: 'Collection not found'
        };
      }
      // ดึง user object ถ้ามี userId
      let user = null;
      let isAdmin = false;
      let username = null;
      if (userId) {
        const userService = require('./userService');
        user = await userService.userService.get_user_by_id(userId);
        if (user) {
          isAdmin = user.role === 'Admin' || user.role === 'SuperAdmin';
          username = user.username;
        }
      }
      // Public collections are accessible to everyone
      if (collection.permission === 'PUBLIC') {
        return {
          success: true,
          data: true
        };
      }
      // Admin/SuperAdmin can access all collections
      if (isAdmin) {
        return {
          success: true,
          data: true
        };
      }
      // Creator has access to all their collections
      if (username && collection.createdBy === username) {
        return {
          success: true,
          data: true
        };
      }
      // Department collections are accessible to department members
      if (collection.permission === 'DEPARTMENT' && user && user.department && collection.department === user.department) {
        return {
          success: true,
          data: true
        };
      }
      // Private: only owner or admin/superadmin
      if (collection.permission === 'PRIVATE') {
        return {
          success: true,
          data: false
        };
      }
      return {
        success: true,
        data: false
      };
    } catch (error) {
      return {
        success: false,
        error: `Failed to check access: ${error instanceof Error ? error.message : 'Unknown error'}`
      };
    }
  }
}

// Export singleton instance
export const collectionService = new CollectionService(); 